### PR TITLE
Replace modular lodash imports with ES2015-style imports

### PR DESCRIPTION
**Pull request by dmsnell**
This PR adds a new codemod to replace our modular lodash imports with the same in ES2015 style.

```js
// input
import _map from 'lodash/map';

// output
import { map as _map } from 'lodash';
```

The reason for making these changes is because we don't need to be importing the modular imports and _if_ we were to remove the Babel transform which magically does this conversion (in reverse) at build-time then we would end up double-including lodash functions since `lodash/map` and `lodash` are separate modules as far as webpack is concerned.

A second codemod introduced here merges all `lodash` imports into a single line where possible. If we have name collisions and need more than one import the "renames" occur as `const` assignments:

```js
// input
import { map } from 'lodash';
import { map as bind } from 'lodash';
import { map as karte } from 'lodash';

// output
import { map } from 'lodash';

// import-renames: auto-generated to prevent name conflicts
// @see bin/codemods/merge-lodash-imports
const bind = map;

const karte = map;
```

**Testing**

Review code changes; smoke test Calypso

### CODE REVIEW

File: client/components/phone-input/phone-number.js [hunk 1]
```diff
@@ -1,11 +1,8 @@
-/**
- * External Dependencies
- */
```
Comment (gwwar) [inline client/components/phone-input/phone-number.js]: We probably want to preserve any comments related to the node. See https://github.com/Automattic/wp-calypso/blob/master/bin/codemods/src/combine-state-utils-imports.js#L50

**Comment by gwwar** [comment]
Looks pretty good so far. Two things we can do in addition:
- [ ] save any external comment blocks
- [ ] update https://github.com/Automattic/wp-calypso/blob/master/bin/codemods/src/combine-state-utils-imports.js to be more generic (basically remove the hard coded `state/utils` and I think that should combine things correctly), and run it against the files here in another commit.

**Comment by aduth** [comment]
>Note: This will result in duplicate import statements from lodash in our modules because it doesn't attempt to merge lodash imports into a single import statement. I will try and write another codemod to follow-up with this work and do just that. I have seen these duplicate entires already before dealing with this codemod

FWIW in a prior codemod I'd worked on for global selectors directory, I'd found a solution to this:

https://github.com/Automattic/wp-calypso/pull/5954#issuecomment-260457603
https://gist.github.com/aduth/904f56aa5594af171d004dd8276b281e

**Comment by dmsnell** [comment]
> FWIW in a prior codemod I'd worked on for global selectors directory, I'd found a solution to this:

thanks! I'm already working on a solution similar to the one you linked. it was good to work through the process and learn, but I'm glad to now have yours to compare against!

File: client/components/phone-input/phone-number.js (hunk 1 above)
Comment (dmsnell) [inline client/components/phone-input/phone-number.js]: Already fixed!

**Comment by dmsnell** [comment]
> run it against the files here in another commit

@gwwar I just made my own `mergeLodashImports` codemod - thanks!

File: bin/codemods/merge-lodash-imports [hunk 2]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env node
+
+/*
+	This codemod converts modular lodash imports to ES2015-style imports
```
Comment (ockham) [inline bin/codemods/merge-lodash-imports]: Comment doesn't apply to this codemod :slightly_smiling_face:

File: client/blocks/credit-card-form/index.jsx [hunk 3]
```diff
@@ -7,17 +7,14 @@ import Gridicon from 'gridicons';
 /**
  * Internal Dependencies
  */
-import camelCase from 'lodash/camelCase';
+import { camelCase, forOwn, kebabCase, mapKeys, values } from 'lodash';
+
```
Comment (ockham) [inline client/blocks/credit-card-form/index.jsx]: Unnecessary newline

File: client/components/happychat/functional.js [hunk 4]
```diff
@@ -1,7 +1,4 @@
-import get from 'lodash/get';
-import { default as collect } from 'lodash/map';
-import find from 'lodash/find';
-import { default as iterate } from 'lodash/forEach';
+import { find, forEach as iterate, get, map as collect } from 'lodash';
```
Comment (ockham) [inline client/components/happychat/functional.js]: Neat!

File: client/components/phone-input/phone-number.js [hunk 5]
```diff
@@ -1,11 +1,8 @@
 /**
  * External Dependencies
  */
-import find from 'lodash/find';
-import map from 'lodash/map';
-import startsWith from 'lodash/startsWith';
-import includes from 'lodash/includes';
-import flatten from 'lodash/flatten';
+import { find, flatten, includes, map, startsWith } from 'lodash';
+
```
Comment (ockham) [inline client/components/phone-input/phone-number.js]: Unnecessary newline

File: client/components/phone-input/index.jsx [hunk 6]
```diff
@@ -1,8 +1,8 @@
 /**
  * External dependencies
  */
-import find from 'lodash/find';
-import identity from 'lodash/identity';
+import { find, identity } from 'lodash';
+
```
Comment (ockham) [inline client/components/phone-input/index.jsx]: Unnecessary newline

File: client/components/plans/plan-price/index.jsx [hunk 7]
```diff
@@ -1,9 +1,9 @@
 /**
  * External dependencies
  */
-import isUndefined from 'lodash/isUndefined';
+import { get, isUndefined } from 'lodash';
+
```
Comment (ockham) [inline client/components/plans/plan-price/index.jsx]: Unnecessary newline

File: client/extensions/woocommerce/state/ui/products/variations/selectors.js [hunk 8]
```diff
@@ -94,5 +94,5 @@ export function getProductVariationsWithLocalEdits( state, productId, siteId = g
 		return { ...variation, ...update };
 	} ) );
 
-	return ( creates || variations ? [ ...creates || [], ...updatedVariations || [] ] : undefined );
+	return creates || variations ? [ ...(creates || []), ...(updatedVariations || []) ] : undefined;
```
Comment (ockham) [inline client/extensions/woocommerce/state/ui/products/variations/selectors.js]: Missing spaces in parens

File: client/lib/posts/post-list-store.js [hunk 9]
```diff
@@ -259,7 +253,7 @@ export default function( id ) {
 		}
 	}
 
-	return new class extends EventEmitter {
+	return new (class extends EventEmitter {
```
Comment (ockham) [inline client/lib/posts/post-list-store.js]: Missing space (or unnecessary paren).

File: client/post-editor/media-modal/index.jsx [hunk 10]
```diff
@@ -51,9 +51,8 @@ function areMediaActionsDisabled( modalView, mediaItems ) {
 		MediaUtils.isItemBeingUploaded( item ) && (
 			// Transients can't be handled by the editor if they are being
 			// uploaded via an external URL
-			! MediaUtils.isTransientPreviewable( item ) ||
-			MediaUtils.getMimePrefix( item ) !== 'image' ||
-			ModalViews.GALLERY === modalView
+			(! MediaUtils.isTransientPreviewable( item ) ||
+			MediaUtils.getMimePrefix( item ) !== 'image' || ModalViews.GALLERY === modalView)
```
Comment (ockham) [inline client/post-editor/media-modal/index.jsx]: Missing spaces in parens

File: client/state/initial-state.js [hunk 11]
```diff
@@ -63,8 +62,8 @@ function addSympathy( initialStateLoader ) {
 	const shouldAdd = (
 		'development' === process.env.NODE_ENV && // only work in local dev mode
 		(
-			Math.random() < 0.25 || // clear 25% of the time
-			config.isEnabled( 'force-sympathy' ) // or whenever the flag is set
+			(// clear 25% of the time
+            (Math.random() < 0.25 || config.isEnabled( 'force-sympathy' ))) // or whenever the flag is set
```
Comment (ockham) [inline client/state/initial-state.js]: Spaces/parens

File: client/state/themes/selectors.js [hunk 12]
```diff
@@ -616,8 +616,7 @@ export function isPremiumThemeAvailable( state, themeId, siteId ) {
 export function isThemeAvailableOnJetpackSite( state, themeId, siteId ) {
 	return !! getTheme( state, siteId, themeId ) || ( // The theme is already available or...
 		isWpcomTheme( state, themeId ) && (  // ...it's a WP.com theme and...
-			config.isEnabled( 'manage/themes/upload' ) &&
-			hasJetpackSiteJetpackThemesExtendedFeatures( state, siteId ) // ...the site supports theme installation from WP.com.
+			(config.isEnabled( 'manage/themes/upload' ) && hasJetpackSiteJetpackThemesExtendedFeatures( state, siteId )) // ...the site supports theme installation from WP.com.
```
Comment (ockham) [inline client/state/themes/selectors.js]: Spaces/parens

**Comment by ockham** [comment]
Noting here that this still leaves us with a bunch of `const stuff = require( 'lodash/stuff' );` style imports. Seems like we should do a codemod sprint some time, starting with the `imports` one 😕

**Comment by ockham** [comment]
Fixed whitespace issues, rebased, and ran codemod on `test/` directory.

**Comment by ockham** [comment]
Had to remove the commit that converted `/test` again since apparently our lodash babel plugin isn't run on that dir.

(This means I messed a bit more with your PR than I intended, sorry about that!)

**Comment by gziolo** [comment]
> Had to remove the commit that converted /test again since apparently our lodash babel plugin isn't run on that dir.

Better skip this dir as we are removing most of the code there with Jest migration :)

File: client/extensions/woocommerce/state/ui/products/variations/selectors.js (hunk 8 above)
Comment (dmsnell) [inline client/extensions/woocommerce/state/ui/products/variations/selectors.js]: I'm pretty sure this came from the code printer. I didn't adjust anything related to parentheses, for example. it was something I was willing to live with in order to clean up the rest, knowing that `prettier` will eventually take care of it.

**Pull request merged**